### ISSUE TITLE
Improper compiler error when checked variable is out of scope

### ISSUE THREAD

**Issue by 2000jedi** [microsoft/checkedc-clang#1017]
# Problem

The following code should compile

```c
int main() {
  int arr _Checked[5];
  int len = 5;
  {
    _Array_ptr<int> dest : count(len) = arr;
  }
  len = 4;
  return 0;
}
```

But instead, the 0.8 (release) version of the compiler generates the following error:

> error: inferred bounds for 'dest' are unknown after statement
> len = 4;

Note that the variable `dest` is already out of scope.

This should not be an error.

**Comment by sulekhark** [comment]
Intersecting the set of in-scope variables across the predecessors of a basic block excludes most variables that are not in-scope for the basic block. However, an exception is when a variable goes out of scope within the basic block. When this happens, the bug noted above manifests.

Fix:
Enabling the AddLifetime flag during CFG construction produces the CFGLifetimeEnds CFGElement for every variable that goes out of scope. When we encounter this CFGElement, we remove the variable that goes out of scope from the ObservedBounds map if it is a checked pointer, and also delete all available expressions that use this out-of-scope variable.

**Comment by sulekhark** [comment]
Fixed by PR #1031.
